File: download_images.py
"""
Download all images from https://loja-oficial-online.shop/ and all linked pages.
Saves images to: loja-oficial/public/images/ preserving directory structure.
"""

import os
import re
import time
import urllib.parse
import urllib.request
from html.parser import HTMLParser
from collections import deque

BASE_URL = "https://loja-oficial-online.shop"
OUTPUT_DIR = r"c:\Users\OlecraM\Downloads\Nova pasta\loja-oficial\public\images"

os.makedirs(OUTPUT_DIR, exist_ok=True)

visited_pages = set()
image_urls = set()

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
}


class LinkImageParser(HTMLParser):
    def __init__(self, base_url):
        super().__init__()
        self.base_url = base_url
        self.links = []
        self.images = []

    def handle_starttag(self, tag, attrs):
        attrs = dict(attrs)
        if tag == "a":
            href = attrs.get("href", "")
            if href:
                full = urllib.parse.urljoin(self.base_url, href)
                if full.startswith(BASE_URL) and "#" not in full:
                    self.links.append(full.rstrip("/"))
        if tag in ("img", "source"):
            for attr in ("src", "srcset", "data-src", "data-lazy-src", "data-srcset"):
                val = attrs.get(attr, "")
                if val:
                    for part in val.split(","):
                        url = part.strip().split(" ")[0]
                        if url:
                            full = urllib.parse.urljoin(self.base_url, url)
                            if full.startswith("http"):
                                self.images.append(full)


def fetch(url):
    req = urllib.request.Request(url, headers=HEADERS)
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            return resp.read().decode("utf-8", errors="ignore")
    except Exception as e:
        print(f"  ERROR fetching {url}: {e}")
        return ""


def url_to_local_path(img_url):
    parsed = urllib.parse.urlparse(img_url)
    # Remove query strings, keep path
    path = parsed.path.lstrip("/")
    # Sanitize
    path = re.sub(r'[<>:"|?*]', "_", path)
    return os.path.join(OUTPUT_DIR, path.replace("/", os.sep))


def download_image(img_url):
    local_path = url_to_local_path(img_url)
    if os.path.exists(local_path):
        return  # already downloaded
    os.makedirs(os.path.dirname(local_path), exist_ok=True)
    try:
        req = urllib.request.Request(img_url, headers=HEADERS)
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = resp.read()
        with open(local_path, "wb") as f:
            f.write(data)
        print(f"  ✓ {img_url}")
    except Exception as e:
        print(f"  ✗ {img_url}: {e}")


def crawl():
    queue = deque([BASE_URL])
    visited_pages.add(BASE_URL)

    while queue:
        page_url = queue.popleft()
        print(f"\n[PAGE] {page_url}")
        html = fetch(page_url)
        if not html:
            continue

        parser = LinkImageParser(page_url)
        parser.feed(html)

        # Collect images
        for img in parser.images:
            if img not in image_urls:
                image_urls.add(img)
                download_image(img)

        # Queue new pages (only same domain, avoid huge crawls)
        for link in parser.links:
            clean = link.split("?")[0]
            if clean not in visited_pages and len(visited_pages) < 200:
                visited_pages.add(clean)
                queue.append(clean)

        time.sleep(0.3)  # be polite

    print(f"\n=== Done! Pages crawled: {len(visited_pages)}, Images downloaded: {len(image_urls)} ===")


if __name__ == "__main__":
    crawl()

File: download-images.mjs
/**
 * download-images.mjs
 * Crawls https://loja-oficial-online.shop/ and downloads ALL images
 * to public/images/ preserving path structure.
 */

import https from "https";
import http from "http";
import fs from "fs";
import path from "path";
import { URL } from "url";

const BASE_URL = "https://loja-oficial-online.shop";
const OUTPUT_DIR = path.join(
  "c:\\Users\\OlecraM\\Downloads\\Nova pasta\\loja-oficial\\public\\images"
);

const visited = new Set();
const downloadedImages = new Set();
const queue = [BASE_URL];
visited.add(BASE_URL);

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0 Safari/537.36",
  Accept: "text/html,application/xhtml+xml,*/*",
};

function fetchText(url) {
  return new Promise((resolve) => {
    const parsed = new URL(url);
    const lib = parsed.protocol === "https:" ? https : http;
    const req = lib.get(
      url,
      { headers: HEADERS, timeout: 20000 },
      (res) => {
        // Follow redirects
        if (
          [301, 302, 303, 307, 308].includes(res.statusCode) &&
          res.headers.location
        ) {
          const next = new URL(res.headers.location, url).href;
          res.resume();
          return fetchText(next).then(resolve);
        }
        let data = "";
        res.setEncoding("utf8");
        res.on("data", (c) => (data += c));
        res.on("end", () => resolve(data));
      }
    );
    req.on("error", () => resolve(""));
    req.on("timeout", () => { req.destroy(); resolve(""); });
  });
}

function fetchBinary(url) {
  return new Promise((resolve) => {
    const parsed = new URL(url);
    const lib = parsed.protocol === "https:" ? https : http;
    const req = lib.get(
      url,
      { headers: HEADERS, timeout: 30000 },
      (res) => {
        if (
          [301, 302, 303, 307, 308].includes(res.statusCode) &&
          res.headers.location
        ) {
          const next = new URL(res.headers.location, url).href;
          res.resume();
          return fetchBinary(next).then(resolve);
        }
        const chunks = [];
        res.on("data", (c) => chunks.push(c));
        res.on("end", () => resolve(Buffer.concat(chunks)));
      }
    );
    req.on("error", () => resolve(null));
    req.on("timeout", () => { req.destroy(); resolve(null); });
  });
}

function extractLinks(html, baseUrl) {
  const links = [];
  // <a href="...">
  const linkRe = /href=["']([^"']+)["']/gi;
  let m;
  while ((m = linkRe.exec(html)) !== null) {
    try {
      const full = new URL(m[1], baseUrl).href.split("?")[0].split("#")[0];
      if (full.startsWith(BASE_URL)) links.push(full);
    } catch {}
  }
  return links;
}

function extractImages(html, baseUrl) {
  const imgs = [];
  // src, srcset, data-src, data-lazy-src, data-srcset, content (og:image)
  const patterns = [
    /(?:src|data-src|data-lazy-src|data-orig-file|data-large-file|data-medium-file)=["']([^"']+)["']/gi,
    /content=["']([^"']+)["']/gi,
  ];
  for (const re of patterns) {
    let m;
    while ((m = re.exec(html)) !== null) {
      const val = m[1];
      // srcset can have multiple entries separated by comma
      for (const part of val.split(",")) {
        const url = part.trim().split(/\s+/)[0];
        if (!url) continue;
        try {
          const full = new URL(url, baseUrl).href.split("?")[0];
          if (/\.(jpg|jpeg|png|gif|webp|svg|avif|ico)(\?|$)/i.test(full)) {
            imgs.push(full);
          }
        } catch {}
      }
    }
  }
  // srcset separately
  const srcsetRe = /srcset=["']([^"']+)["']/gi;
  let sm;
  while ((sm = srcsetRe.exec(html)) !== null) {
    for (const part of sm[1].split(",")) {
      const url = part.trim().split(/\s+/)[0];
      if (!url) continue;
      try {
        const full = new URL(url, baseUrl).href.split("?")[0];
        if (/\.(jpg|jpeg|png|gif|webp|svg|avif|ico)(\?|$)/i.test(full)) {
          imgs.push(full);
        }
      } catch {}
    }
  }
  return imgs;
}

function urlToLocalPath(imgUrl) {
  const parsed = new URL(imgUrl);
  const p = parsed.pathname.replace(/^\//, "");
  return path.join(OUTPUT_DIR, ...p.split("/"));
}

async function downloadImage(imgUrl) {
  if (downloadedImages.has(imgUrl)) return;
  downloadedImages.add(imgUrl);

  const localPath = urlToLocalPath(imgUrl);
  if (fs.existsSync(localPath)) {
    console.log(`  (exists) ${imgUrl}`);
    return;
  }

  fs.mkdirSync(path.dirname(localPath), { recursive: true });
  const data = await fetchBinary(imgUrl);
  if (!data || data.length === 0) {
    console.log(`  ✗ ${imgUrl}`);
    return;
  }
  fs.writeFileSync(localPath, data);
  console.log(`  ✓ ${path.basename(localPath)} (${(data.length / 1024).toFixed(1)} KB)`);
}

async function crawl() {
  console.log(`Downloading images from ${BASE_URL}`);
  console.log(`Saving to: ${OUTPUT_DIR}\n`);

  while (queue.length > 0 && visited.size <= 300) {
    const pageUrl = queue.shift();
    console.log(`\n[PAGE ${visited.size}] ${pageUrl}`);

    const html = await fetchText(pageUrl);
    if (!html) continue;

    // Download images on this page
    const imgs = extractImages(html, pageUrl);
    for (const img of imgs) {
      await downloadImage(img);
    }

    // Add new pages to crawl
    const links = extractLinks(html, pageUrl);
    for (const link of links) {
      if (!visited.has(link) && visited.size < 300) {
        visited.add(link);
        queue.push(link);
      }
    }

    await new Promise((r) => setTimeout(r, 200)); // polite delay
  }

  console.log(`\n=== DONE ===`);
  console.log(`Pages crawled: ${visited.size}`);
  console.log(`Images downloaded: ${downloadedImages.size}`);
  console.log(`Saved to: ${OUTPUT_DIR}`);
}

crawl().catch(console.error);
